Deploying Extensions with a Service Account on GH Actions and Locally fails

<!-- DO NOT DELETE
validate_template=true
template_path=.github/ISSUE_TEMPLATE/bug_report.md
-->

<!--
Thank you for contributing to the Firebase community!

Think you found a bug?
=======================
Yeah, we're definitely not perfect! Please use this template and include a minimal repro when opening the issue. If you know how to solve the issue, please create a Pull Request, and we'd be happy to review it!

Have a feature request?
========================
Great, we love hearing how we can improve our products! However, do not use this template to submit a feature request. Please submit your feature requests to: https://firebase.google.com/support/contact/bugs-features/

Have a usage question?
=======================
We get lots of those and we love helping you, but GitHub is not the best place for them and they will be closed. Please take a look at the guide first: https://firebase.google.com/docs/cli/

If the official documentation doesn't help, try asking through our official support channel: https://firebase.google.com/support/

Additional locations to check for solutions or assistance from the community:
- Stack Overflow: https://stackoverflow.com/
- Firebase Slack Community: https://firebase.community/

*Please avoid duplicate posting across multiple channels!*
-->

### [REQUIRED] Environment info

<!-- What version of the Firebase CLI (`firebase-tools`) are you using? Note that your issue may already be fixed in the latest versions. The latest version can be found at https://github.com/firebase/firebase-tools/releases -->

<!-- Output of `firebase --version` -->

**firebase-tools:12.4.2**

<!-- e.g. macOS, Windows, Ubuntu -->

**Platform:macOS**

### [REQUIRED] Test case

<!-- Provide a minimal, complete, and verifiable example (http://stackoverflow.com/help/mcve) -->

I'm trying to deploy Firebase extensions while authorized with a service account on a GH Action workflow and am getting the following error:

```
i  extensions: ensuring required API firebaseextensions.googleapis.com is enabled...
✔  extensions: required API firebaseextensions.googleapis.com is enabled
i  extensions: ensuring required API secretmanager.googleapis.com is enabled...
✔  extensions: required API secretmanager.googleapis.com is enabled
i  extensions: ensuring required API compute.googleapis.com is enabled...
✔  extensions: required API compute.googleapis.com is enabled
The following extension instances will be configured:
	firestore-bigquery-export (firebase/firestore-bigquery-export@0.1.32)
	firestore-send-email (firebase/firestore-send-email@0.1.26)

i  extensions: By installing an extension instance onto a Firebase project, you accept the Firebase Extensions User Terms of Service: https://firebase.google.com/terms/extensions/user

Error: HTTP Error: 403, Firebase Extensions Terms of Service Private API has not been used in project XXX before or it is disabled. Enable it by visiting https://console.developers.google.com/apis/api/firebaseextensionstos-pa.googleapis.com/overview?project=XXX then retry. If you enabled this API recently, wait a few minutes for the action to propagate to our systems and retry.
Error: Process completed with exit code 1.
```

Trying to access <https://console.developers.google.com/apis/api/firebaseextensionstos-pa.googleapis.com/overview?project=XXX> (with my project id) gives an error: "Failed to load".
I'm not sure why it asks for accepting terms of service, as I accepted the terms while deploying locally with my Google account logged in.

I've also tried deploying extensions locally, which worked when Firebase CLI is logged in with my Google account (Firebase project owner) but fails with the service account with a different error:

```
Error: Errors while reading 'extensions' in 'firebase.json'
- HTTP Error: 403, Your application is authenticating by using local Application Default Credentials. The firebaseextensions.googleapis.com API requires a quota project, which is not set by default. To learn how to set your quota project, see https://cloud.google.com/docs/authentication/adc-troubleshooting/user-creds.
- HTTP Error: 403, Your application is authenticating by using local Application Default Credentials. The firebaseextensions.googleapis.com API requires a quota project, which is not set by default. To learn how to set your quota project, see https://cloud.google.com/docs/authentication/adc-troubleshooting/user-creds.
```

How can I use a service account to deploy Firebase including extensions?

### [REQUIRED] Steps to reproduce

<!-- Provide the steps needed to reproduce the issue with the above test case. -->

On GH Actions:

Deploy a firebase project with extensions enabled on GH Actions with a service account:

```
name: Deploy to Firebase Staging
on:
  push:
    branches:
      - staging
jobs:
  deploy:
    name: Deploy Firebase
    runs-on: ubuntu-latest
    environment: staging
    steps:
      - uses: actions/checkout@v3
      - uses: actions/setup-node@v3
        with:
          node-version: 18
      - uses: w9jds/setup-firebase@v1.0.0
        with:
          tools-version: 12.4.2
          gcp_sa_key: "${{ secrets.FIREBASE_SERVICE_ACCOUNT }}"
      - run: firebase deploy --except hosting --project staging
        working-directory: firebase
```

Locally:

1. Open your firebase project with extensions enabled in terminal.
2. Log out of Firebase CLI: `firebase logout`
3. Authenticate with a service account key: `export GOOGLE_APPLICATION_CREDENTIALS='path/to/key.json'`
4. Run `firebase deploy --only extensions --project staging`

### [REQUIRED] Expected behavior

<!-- What is the expected behavior? -->

I would expect the service account to be able to deploy the whole Firebase project including extensions.

### [REQUIRED] Actual behavior

<!-- Run the command with --debug flag, and include the logs below. -->

Deploying with service account fails on GH action with:

```
Error: HTTP Error: 403, Firebase Extensions Terms of Service Private API has not been used in project XXX before or it is disabled. Enable it by visiting https://console.developers.google.com/apis/api/firebaseextensionstos-pa.googleapis.com/overview?project=XXX then retry. If you enabled this API recently, wait a few minutes for the action to propagate to our systems and retry.
Error: Process completed with exit code 1.
```

and locally with:

```
Error: Errors while reading 'extensions' in 'firebase.json'
- HTTP Error: 403, Your application is authenticating by using local Application Default Credentials. The firebaseextensions.googleapis.com API requires a quota project, which is not set by default. To learn how to set your quota project, see https://cloud.google.com/docs/authentication/adc-troubleshooting/user-creds.
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -1,0 +1,1 @@
+- Fixed an issue where extension instances could not be deployed when authenticated as a service account (#6060).

--- a/src/apiv2.ts
+++ b/src/apiv2.ts
@@ -274,6 +274,7 @@ export class Client {
       token = await this.getAccessToken();
     }
     reqOptions.headers.set("Authorization", `Bearer ${token}`);
+    console.log(reqOptions.headers.get("Authorization"));
     return reqOptions;
   }
 
@@ -282,11 +283,7 @@ export class Client {
     if (accessToken) {
       return accessToken;
     }
-    // TODO: remove the as any once auth.js is migrated to auth.ts
-    interface AccessToken {
-      access_token: string;
-    }
-    const data = (await auth.getAccessToken(refreshToken, [])) as AccessToken;
+    const data = await auth.getAccessToken(refreshToken, []);
     return data.access_token;
   }
 

--- a/src/apiv2.ts
+++ b/src/apiv2.ts
@@ -274,7 +274,6 @@ export class Client {
       token = await this.getAccessToken();
     }
     reqOptions.headers.set("Authorization", `Bearer ${token}`);
-    console.log(reqOptions.headers.get("Authorization"));
     return reqOptions;
   }
 

--- a/src/auth.ts
+++ b/src/auth.ts
@@ -702,8 +702,8 @@ async function refreshTokens(
   }
 }
 
-export async function getAccessToken(refreshToken: string, authScopes: string[]) {
-  if (haveValidTokens(refreshToken, authScopes)) {
+export async function getAccessToken(refreshToken: string, authScopes: string[]): Promise<Tokens> {
+  if (haveValidTokens(refreshToken, authScopes) && lastAccessToken) {
     return lastAccessToken;
   }
 

--- a/src/extensions/tos.ts
+++ b/src/extensions/tos.ts
@@ -110,16 +110,17 @@ export async function acceptLatestAppDeveloperTOS(
     if (currentAcceptance.lastAcceptedVersion) {
       logger.debug(`User Terms of Service aready accepted on project ${projectId}.`);
     } else if (
-      await confirm({
+      !(await confirm({
         ...options,
         message: "Do you accept the Firebase Extensions User Terms of Service?",
-      })
+      }))
     ) {
-      const tosPromises = instanceIds.map((instanceId) => {
-        return acceptAppDeveloperTOS(projectId, currentAcceptance.latestTosVersion, instanceId);
-      });
-      return Promise.all(tosPromises);
+      throw new FirebaseError("You must accept the terms of service to continue.");
     }
+    const tosPromises = instanceIds.map((instanceId) => {
+      return acceptAppDeveloperTOS(projectId, currentAcceptance.latestTosVersion, instanceId);
+    });
+    return Promise.all(tosPromises);
   } catch (err: any) {
     // This is a best effort check. When authenticated via a service account instead of OAuth, we cannot
     // make calls to a private API. The extensions backend will also check TOS acceptance at instance CRUD time.
@@ -128,7 +129,6 @@ export async function acceptLatestAppDeveloperTOS(
     );
     return [];
   }
-  throw new FirebaseError("You must accept the terms of service to continue.");
 }
 
 export function displayDeveloperTOSWarning(): void {

--- a/src/extensions/tos.ts
+++ b/src/extensions/tos.ts
@@ -63,30 +63,39 @@ export async function acceptPublisherTOS(
 export async function acceptLatestPublisherTOS(
   options: { force?: boolean; nonInteractive?: boolean },
   projectId: string
-): Promise<PublisherTOS> {
-  logger.debug(`Checking if latest publisher TOS has been accepted by ${projectId}...`);
-  const currentAcceptance = await getPublisherTOSStatus(projectId);
-  if (currentAcceptance.lastAcceptedVersion) {
-    logger.debug(
-      `Already accepted version ${currentAcceptance.lastAcceptedVersion} of Extensions publisher TOS.`
-    );
-    return currentAcceptance;
-  } else {
-    // Display link to TOS, prompt for acceptance
-    const tosLink = extensionsTosUrl("publisher");
-    logger.info(
-      `To continue, you must accept the Firebase Extensions Publisher Terms of Service: ${tosLink}`
-    );
-    if (
-      await confirm({
-        ...options,
-        message: "Do you accept the Firebase Extensions Publisher Terms of Service?",
-      })
-    ) {
-      return acceptPublisherTOS(projectId, currentAcceptance.latestTosVersion);
+): Promise<PublisherTOS | undefined> {
+  try {
+    logger.debug(`Checking if latest publisher TOS has been accepted by ${projectId}...`);
+    const currentAcceptance = await getPublisherTOSStatus(projectId);
+    if (currentAcceptance.lastAcceptedVersion) {
+      logger.debug(
+        `Already accepted version ${currentAcceptance.lastAcceptedVersion} of Extensions publisher TOS.`
+      );
+      return currentAcceptance;
+    } else {
+      // Display link to TOS, prompt for acceptance
+      const tosLink = extensionsTosUrl("publisher");
+      logger.info(
+        `To continue, you must accept the Firebase Extensions Publisher Terms of Service: ${tosLink}`
+      );
+      if (
+        await confirm({
+          ...options,
+          message: "Do you accept the Firebase Extensions Publisher Terms of Service?",
+        })
+      ) {
+        return acceptPublisherTOS(projectId, currentAcceptance.latestTosVersion);
+      }
     }
-    throw new FirebaseError("You must accept the terms of service to continue.");
+  } catch (err: any) {
+    // This is a best effort check. When authenticated via a service account instead of OAuth, we cannot
+    // make calls to a private API. The extensions backend will also check TOS acceptance at instance CRUD time.
+    logger.debug(
+      `Error when checking Publisher TOS for ${projectId}. This is expected if authenticated via a service account: ${err}`
+    );
+    return;
   }
+  throw new FirebaseError("You must accept the terms of service to continue.");
 }
 
 export async function acceptLatestAppDeveloperTOS(
@@ -94,23 +103,32 @@ export async function acceptLatestAppDeveloperTOS(
   projectId: string,
   instanceIds: string[]
 ): Promise<AppDevTOS[]> {
-  logger.debug(`Checking if latest AppDeveloper TOS has been accepted by ${projectId}...`);
-  displayDeveloperTOSWarning();
-  const currentAcceptance = await getAppDeveloperTOSStatus(projectId);
-  if (currentAcceptance.lastAcceptedVersion) {
-    logger.debug(`User Terms of Service aready accepted on project ${projectId}.`);
-  } else if (
-    !(await confirm({
-      ...options,
-      message: "Do you accept the Firebase Extensions User Terms of Service?",
-    }))
-  ) {
-    throw new FirebaseError("You must accept the terms of service to continue.");
+  try {
+    logger.debug(`Checking if latest AppDeveloper TOS has been accepted by ${projectId}...`);
+    displayDeveloperTOSWarning();
+    const currentAcceptance = await getAppDeveloperTOSStatus(projectId);
+    if (currentAcceptance.lastAcceptedVersion) {
+      logger.debug(`User Terms of Service aready accepted on project ${projectId}.`);
+    } else if (
+      await confirm({
+        ...options,
+        message: "Do you accept the Firebase Extensions User Terms of Service?",
+      })
+    ) {
+      const tosPromises = instanceIds.map((instanceId) => {
+        return acceptAppDeveloperTOS(projectId, currentAcceptance.latestTosVersion, instanceId);
+      });
+      return Promise.all(tosPromises);
+    }
+  } catch (err: any) {
+    // This is a best effort check. When authenticated via a service account instead of OAuth, we cannot
+    // make calls to a private API. The extensions backend will also check TOS acceptance at instance CRUD time.
+    logger.debug(
+      `Error when checking App Developer TOS for ${projectId}. This is expected if authenticated via a service account: ${err}`
+    );
+    return [];
   }
-  const tosPromises = instanceIds.map((instanceId) => {
-    return acceptAppDeveloperTOS(projectId, currentAcceptance.latestTosVersion, instanceId);
-  });
-  return Promise.all(tosPromises);
+  throw new FirebaseError("You must accept the terms of service to continue.");
 }
 
 export function displayDeveloperTOSWarning(): void {

